kernel/uVars.ml

```diff
@@ -285,36 +288,36 @@ struct
 
   let make names (univs, _ as x) : t =
     let qs, us = Instance.to_array univs in
-    assert (Array.length (fst names) = Array.length qs && Array.length(snd names) = Array.length us);
+    assert (Array.length names.quals = Array.length qs && Array.length names.univs = Array.length us);
     (names, x)
 
   (** Universe contexts (variables as a list) *)
-  let empty = (([||], [||]), (Instance.empty, Constraints.empty))
+  let empty = (empty_bound_names, (Instance.empty, Constraints.empty))
   let is_empty (_, (univs, csts)) = Instance.is_empty univs && Constraints.is_empty csts
 
   let pr prq prl ?variance (_, (univs, csts) as uctx) =
     if is_empty uctx then mt() else
       h (Instance.pr prq prl ?variance univs ++ str " |= ") ++ h (v 0 (Constraints.pr prl csts))
 
-  let hcons ((qnames, unames), (univs, csts)) =
+  let hcons ({quals = qnames; univs = unames}, (univs, csts)) =
     let hqnames, qnames = Hashcons.hashcons_array Names.Name.hcons qnames in
     let hunames, unames = Hashcons.hashcons_array Names.Name.hcons unames in
     let hunivs, univs = Instance.hcons univs in
     let hcsts, csts = Constraints.hcons csts in
-    Hashset.Combine.combine4 hqnames hunames hunivs hcsts, ((qnames, unames), (univs, csts))
+    Hashset.Combine.combine4 hqnames hunames hunivs hcsts, ({quals = qnames; univs = unames}, (univs, csts))
 
   let names ((names, _) : t) = names
   let instance (_, (univs, _csts)) = univs
   let constraints (_, (_univs, csts)) = csts
 
-  let union ((qna, una), (univs, csts)) ((qna', una'), (univs', csts')) =
-    (Array.append qna qna', Array.append una una'), (Instance.append univs univs', Constraints.union csts csts')
+  let union (names, (univs, csts)) (names', (univs', csts')) =
+    { quals = Array.append names.quals names'.quals; univs =  Array.append names.univs names'.univs}, (Instance.append univs univs', Constraints.union csts csts')
```

Comment:
maybe worth defining a `append_bound_names` helper, this happens in several places